### PR TITLE
CSOLabelRenderer Tutorial

### PR DESCRIPTION
**Pull request by SamaaAhm**

### CODE REVIEW

**Review by sahehb — CHANGES_REQUESTED** [review]
see comments

**Review by sahehb — APPROVED** [review]

File: mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md [hunk 1]
```diff
@@ -16,112 +16,216 @@ menu:
...
-if cso.getSubType() == 'distanceLine':
-    labelString = 'Length: ' + str(cso.getLength())
-    labelName = 'Distance '
-else:
-    labelString = ''
-    labelName = ''
-labelName += str(cso.getId())
-labelCaption = labelName + ":  "
-
+labelString = 'Length: ' + str(round(cso.getLength(), 2)) + ' mm'
```
Comment (mrbean-bremen) [inline mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md:62]: Same here:
```py
labelString = f'Length: {cso.getLength():.2f} mm'
```

File: mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md [hunk 2]
```diff
@@ -16,112 +16,216 @@ menu:
...
+In order to see all possible parameters of a CSO, add a `CSOInfo` module to your network and connect it to the `CSOManager`. The geometric informations of the selected CSO from `CSOManager` can be seen there.
 
-![Labeled DistanceLine in View2D](/images/tutorials/dataobjects/contours/Ex6_11.png "Labeled DistanceLine in View2D")
+![CSOInfo](/images/tutorials/dataobjects/contours/Ex6_CSOInfo.png "CSOInfo")
 
-In the `CSOLabelRenderer` panel, head to the *Options* tab. Here, you can fine-tune various settings to enhance the appearance of CSO labels. Set the *Label Border Margin X* to *5* to create a margin around selected CSOs. Increase the *Font Size* to *15* for larger and more readable text. Additionally, adjust the *Min Connecting Line Length* to *15* to specify a minimum distance for drawing connecting lines, maintaining a clean and organized layout. 
+For labels shown on greyscale images, it makes sense to add a shadow to your labels. Open the panel of the `SoCSOVisualizationSettings` module and on tab *Misc* check option *Should render shadow*. This increases readability of your labels.
```
Comment (mrbean-bremen) [inline mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md:70]: remove "to your labels"
check option -> check the option
increases readability -> increases the readability

File: mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md [hunk 3]
```diff
@@ -16,112 +16,216 @@ menu:
...
+If you want to define your static text as a parameter in multiple labels, you can open the panel of the `CSOLabelRenderer` module and define text as User Data. The values can then be used in Python via *userData*.
 
-![SoCSORectangleEditor](/images/tutorials/dataobjects/contours/Ex6_7.png "SoCSORectangleEditor")
+![User Data](/images/tutorials/dataobjects/contours/Ex6_Parameters.png "User Data")
 
-Ensure to update the *Creator Extension Id* field in the `SoView2DCSOExtensibleEditor` module with the *Extension Id* of the `SoCSORectangleEditor` module.
+You can also add multiple CSO editors to see the different options. Add the `SoCSORectangleEditor` module to your workspace and connect it to the `SoGroup` module. As we now have two different editors, we need to tell the `CSOLabelRenderer` which CSO to be rendered how. Open the panel of the `SoCSODistanceLineEditor`. You can see the field *Extension Id* set to *distanceLine*. Open the panel of the `SoCSORectangleEditor`. You can see the field *Extension Id* set to *rectangle*.
```
Comment (mrbean-bremen) [inline mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md:78]: which CSO to be rendered -> which CSO is to be rendered (or: will be rendered)

File: mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md [hunk 4]
```diff
@@ -16,112 +16,216 @@ menu:
...
 
-Open the `CSOLabelRenderer` panel and add similar script as for the *distanceLine* subtype, but include *Area* calculation as well, since we are dealing with a rectangle.
+We currently defined the *labelName* and *labelString* for the distance line. If we want to define different labels for different types of CSOs, we have to change the `CSOLabelRenderer` Python script. Open the panel of the `CSOLabelRenderer` and change the Python code to the following:
 
-{{< highlight filename="" >}}
+{{< highlight filename="CSOLabelRenderer" >}}
 ```Python
-...
+if cso.getSubType() == 'distanceLine':
+  labelString = userData0 + str(round(cso.getLength(), 2)) + ' mm'
```
Comment (mrbean-bremen) [inline mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md:87]: see above

File: mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md [hunk 5]
```diff
@@ -16,112 +16,216 @@ menu:
...
+  labelName = userData1
+  labelName += str(cso.getId())
 elif cso.getSubType() == 'rectangle':
-    labelString = str('Length: ') + str(cso.getLength()) + '\n'
-    labelString += str('Area: ') + str(cso.getArea()) + '\n'
-    labelName = 'Rectangle '
-...
+  labelString = userData0 + str(round(cso.getLength(), 2)) + ' mm\n'
+  labelString += userData2 + str(round(cso.getArea(), 2)) + ' mm^2'
```
Comment (mrbean-bremen) [inline mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md:92]: same

File: mevislab.github.io/content/tutorials/shorts.md [hunk 6]
```diff
@@ -157,3 +158,21 @@ QApplication.setPalette(palette)
 This script defines the color of the MeVisLab user interface elements. You can define other colors and more items, this is just an example of what you can do with user scripts.
 
 Switch back to the MeVisLab IDE and select the menu item {{< menuitem "Extras" "Reload Module Database (Clear Cache)">}} again. The colors of the MeVisLab IDE change as defined in our Python script. This change persists until you restart MeVisLab and can always be repeated by selecting the menu entry or the keyboard shortcut {{< keyboard "ctrl+F9" >}}.
+
+## Show status of module in- and output {#mlimagestate}
+
+Especially in large networks it is useful to see the state of the in- and output connectors of a module. By default, the module connectors do not show if data is available. Below image shows a `DicomImport` module and a `View2D` module where no data is loaded.
```
Comment (mrbean-bremen) [inline mevislab.github.io/content/tutorials/shorts.md:164]: input and output

File: mevislab.github.io/content/tutorials/shorts.md [hunk 7]
```diff
@@ -157,3 +158,21 @@ QApplication.setPalette(palette)
...
+
+![No status on connector](/images/tutorials/LMIMageState_Off.png "No status on connector")
+
+In MeVisLab preferences dialog, you can see a checkbox *Show ML image state*. By default, the setting is *Off*.
```
Comment (mrbean-bremen) [inline mevislab.github.io/content/tutorials/shorts.md:168]: In the ...

File: mevislab.github.io/content/tutorials/shorts.md [hunk 8]
```diff
@@ -157,3 +158,21 @@ QApplication.setPalette(palette)
...
+
+![Show ML image state](/images/tutorials/LMIMageState.png "Show ML image state")
+
+After enabling *Show ML image state*, your network changes and the in- and output connectors appear red in case no data is available at the output.
```
Comment (mrbean-bremen) [inline mevislab.github.io/content/tutorials/shorts.md:172]: input and output

File: mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md [hunk 9]
```diff
@@ -45,7 +45,7 @@ Enter the following to the panel of the `CSOLabelRenderer` module:
 {{< highlight filename="CSOLabelRenderer" >}}
 ```Python
 labelString = f"Length {cso.getLength()}"  
-labelName = f"Distance: {cso.getID:.2f}"
+labelName = f"Distance: {cso.getId:.2f}"
```
Comment (mrbean-bremen) [inline mevislab.github.io/content/tutorials/dataobjects/contours/contourexample6.md:48]: It's `getId()` (wth the parenthesis)

**Review by mrbean-bremen — APPROVED** [review]

**Pull request merged**